docs/proposals/operator-custom-crd-download-url/README.md

```diff
@@ -0,0 +1,103 @@
+specify the version of the control plane components. In that scenario, this assumption breaks, and we encounter the risk of the operator installing the wrong CRDs for that Karmada instance.
+
+The default behavior is sensible and will be retained. However, it should not be the behavior when the intent is to create a new Karmada instance with control plane components of a different version than that of the operator's release version. 
+To support that scenario, we require the ability to explicitly set, as part of the Karmada resource, the URL that should be used to download the CRDs tarball for that Karmada instance.
+
+In addition to making sure required CRDs are downloaded, the Karmada operator also employs a caching strategy to avoid re-downloading CRDs it has already downloaded. It's a cache on the local file system.
+Downloaded CRDs are stored at a well-known location on the local file system based the release version of the Karmada operator. When provisioning a new Karmada instance, before attempting to download the required CRDs, the operator will fist check if
+they already exist in the cache. To support custom CRD download URLs, we require this caching strategy be extended to be more generic.
+
+## Design Details
+
+To support downloading CRDs from a custom URL, we need to add the new fields `CRDRemoteURL` and `CRDDownloadPolicy` to the Karmada resource spec:
+```go
+// CRDDownloadPolicy specifies a policy for how the Karmada operator will retrieve the Karmada CRDs when provisioning a Karmada instance.
+// +enum
+type CRDDownloadPolicy string
+
+const (
+    // DownloadAlways instructs the Karmada operator to always download the CRDs tarball from a remote location.
+    DownloadAlways CRDDownloadPolicy = "Always"
+    // DownloadIfNotPresent instructs the Karmada operator to download the CRDs tarball from a remote location only if it is not yet present in the local cache.
+    DownloadIfNotPresent CRDDownloadPolicy = "IfNotPresent"
+)
+
+// KarmadaSpec is the specification of the desired behavior of the Karmada.
+type KarmadaSpec struct {
+
+HostCluster *HostCluster `json:"hostCluster,omitempty"`
+
+PrivateRegistry *ImageRegistry `json:"privateRegistry,omitempty"`
+
+Components *KarmadaComponents `json:"components,omitempty"`
+
+FeatureGates map[string]bool `json:"featureGates,omitempty"`
+
+// Custom URL that should be used to download required CRDs for this Karmada instance.
+// Omit if the intent is to provision a new Karmada instance with control plane components of the same version as the operator's release version.
+// In that scenario, the CRDs tarball will be downloaded from a release artifact on GitHub that's of the same version as the operator's release version.
+// When that is not the intent, this field should be explicitly set to ensure the right version of the CRDs is installed for this Karmada instance.
+// +optional
+CRDRemoteURL string `json:"crdRemoteURL,omitempty"`
```

Comment:
What if another user stores the tarball on a FTP server, is it a reasonable case? can we support that case?

Can we get some inspiration from Kubernetes [ProbeHandler](https://github.com/kubernetes/kubernetes/blob/4a0a7e809237e74afd4ea1499520063f96158e1e/staging/src/k8s.io/api/core/v1/types.go#L2768-L2784) which can supports HTTP, TCP, GRPC, and so on. The `ProbeHandler` is highly extendable, it can add more actions without effect the previous users.

What do you think?

Reply:
Given the ubiquitous use of HTTP/HTTPS for accessing artifacts, I find it very hard to think of scenarios where using FTP/SFTP would be preferred. However, that could be because I'm missing some context. I do agree with you that we should leave the door open to support that use case down the line if the need for it arises. I think implementing it now though is out of scope for this feature, but I'm happy you brought it up. This proposed design is flexible, so if we do decide to support other protocols like FTP/SFTP down the line, it's possible without changes to the API.
For instance, `CRDRemoteURL` can just be specified like so: `sftp://my-server.com/karmada/artifacts/v1.10.2/crds.tar.gz`. Then at runtime, based the the URL's scheme, we can decide which strategy should be used to download the CRDs, and leverage a pattern like the [ProbeHandler](https://github.com/kubernetes/kubernetes/blob/4a0a7e809237e74afd4ea1499520063f96158e1e/staging/src/k8s.io/api/core/v1/types.go#L2768-L2784) as you suggested to implement the logic.

Reply:
> I think implementing it now though is out of scope for this feature

Yes, exactly. I don't expect to support other protocols in this proposal. Just make room for future extensions, without breaking changes and being ambitious.  Let me see if I can help with it.

By the way, can you help to share more details about your HTTPServer where servers the CRD tarball? Does it support both HTTP and HTTPS? If it verifies the client's certificate?

Reply:
Hi Joe, Please help to look at how this look like:
```golang
// KarmadaSpec is the specification of the desired behavior of the Karmada.
type KarmadaSpec struct {
	// otherfields

	// CRDTarball defines the source location from where the CRD tarball should be downloaded,
	// along with the download policy.
	// If not set, the Karmada Operator will download the CRDs from the GitHub release.
	// By default, it will fetch the tarball corresponding to the version of the operator itself.
	// For instance, if the operator's version is v1.10.0, it will download from:
	// https://github.com/karmada-io/karmada/releases/download/v1.10.0/crds.tar.gz
	// Note that the download will only occur if the tarball is not already present in the local cache.
	// +optional
	CRDTarball *CRDTarball `json:"crdTarball,omitempty"`
}


// CRDDownloadPolicy specifies a policy for how the Karmada operator will retrieve
// the Karmada CRDs when provisioning a Karmada instance.
type CRDDownloadPolicy string

const (
	// DownloadAlways instructs the Karmada operator to always download the CRDs tarball from a remote location.
	DownloadAlways CRDDownloadPolicy = "Always"

	// DownloadIfNotPresent instructs the Karmada operator to download the CRDs tarball from a remote location
	// only if it is not yet present in the local cache.
	DownloadIfNotPresent CRDDownloadPolicy = "IfNotPresent"
)

// CRDTarball defines the source from which the CRD tarball can be downloaded and the download policy.
type CRDTarball struct {

	// HTTPSource is the full URL for downloading the CRD tarball using either HTTP or HTTPS protocol.
	// +optional
	HTTPSource *string `json:"httpSource,omitempty"`

	// NOTE: We still can add more fields like TLS certificate if there is a need in the future.

	// NOTE: We can add more protocol support here. But this will not be included this time.
	// FTPSource is the full URL for downloading the CRD tarball using FTP protocol.
	// If specified, HTTPSource must not be set.
	// +optional
	// FTPSource *string `json:"ftpServer,omitempty"`

	// DownloadPolicy specifies a policy that should be used by the Karmada operator to retrieve
	// the Karmada CRDs when provisioning this Karmada instance.
	// Valid values are "Always" and "IfNotPresent".
	// Defaults to "IfNotPresent".
	// +kubebuilder:validation:Enum=Always;IfNotPresent
	// +kubebuilder:default=IfNotPresent
	// +optional
	DownloadPolicy *CRDDownloadPolicy `json:"downloadPolicy,omitempty"`
}
```

Reply:
Thanks for clarifying. This structure does provide a great point of extensibility.

Reply:
Can use either HTTP OR HTTPS. No specific cert requirements.

Reply:
Will push update shortly to address comments.